### PR TITLE
fix: multiple reconciliation of same event in FRM

### PR DESCRIPTION
**Pull request by tanishagoyal2**
## Type of Change
- [x] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [ ] Documentation/CI
- [x] Fault Management
- [ ] Health Monitors
- [ ] Janitor
- [ ] Other: ____________

## Testing
- [x] Tests pass locally
- [x] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [x] Self-review completed
- [x] Documentation updated (if needed)
- [x] Ready for review

##Summary
This PR targets few bug fixes in fault-remediation module. 
1. FRM receiving same event multiple times with diff status values: Updated postgres and mongodb pipeline in BuildQuarantinedAndDrainedNodesPipeline to check if the update has been done in expected field or not 
2. fault-quarantine was not adding label `dgxc.nvidia.com/nvsentinel-state` for taint only config: updated the code to add this label for both taint only, cordon only and both config.
Screenshot 2026-02-23 at 12 07 56 PM

we can see the cordon label and taint but not nvsentinel state label

3. the userpodsevictionstatus is not updated correctly for health event: It was showing duplicate fields

Screenshot 2026-02-23 at 7 55 48 PM

we can see the event with duplicate fields like status/Status and message/Message

4. The event updated by FRM didn't have lastremediationtimestamp field updated: the postgres query was not updating the lastremediationtimestamp field. 

Screenshot 2026-02-24 at 2 59 15 PM

We can see that the event is updated faultRemediated=true but we are missing lastremediationtimestamp in event

## Summary by CodeRabbit

* **New Features**
  * Enhanced event detection for quarantine and drain operations with explicit branching logic for UPDATE and INSERT scenarios.

* **Bug Fixes**
  * Refined sentinel state label assignment to require taints when setting quarantined status.
  * Improved NULL state handling for node quarantine status updates.

* **Chores**
  * Minor code cleanup and structural optimization.

### CODE REVIEW

File: store-client/pkg/client/postgresql_pipeline_builder.go [hunk 1]
```diff
@@ -192,21 +192,28 @@ func (b *PostgreSQLPipelineBuilder) BuildQuarantinedAndDrainedNodesPipeline() da
...
+										datastore.E("$in", datastore.A(string(model.StatusSucceeded), string(model.AlreadyDrained))),
+									)),
+								)),
 								datastore.E("fullDocument.healtheventstatus.userpodsevictionstatus.status", datastore.D(
 									datastore.E("$in", datastore.A(string(model.StatusSucceeded), string(model.AlreadyDrained))),
 								)),
-								datastore.E("fullDocument.healtheventstatus.nodequarantined", datastore.D(
-									datastore.E("$in", datastore.A(string(model.Quarantined), string(model.AlreadyQuarantined))),
-								)),
 							),
```
[the remark this answers is not in the document]
Comment (tanishagoyal2) [reply store-client/pkg/client/postgresql_pipeline_builder.go:203]: I have updated the code. please check now

File: store-client/pkg/client/mongodb_pipeline_builder.go [hunk 2]
```diff
@@ -144,34 +144,102 @@ func (b *MongoDBPipelineBuilder) BuildProcessableNonFatalUnhealthyInsertsPipelin
...
+								datastore.E("fullDocument.healtheventstatus.nodequarantined", string(model.UnQuarantined)),
+								datastore.E("fullDocument.healtheventstatus.userpodsevictionstatus.status", string(model.StatusSucceeded)),
+							),
+							// Watch for cancelled quarantine events (for annotation cleanup)
+							datastore.D(
+								datastore.E("updateDescription.updatedFields", datastore.D(
+									datastore.E("healtheventstatus.nodequarantined", string(model.Cancelled)),
+								)),
+							),
 						)),
```
[the remark this answers is not in the document]
Comment (tanishagoyal2) [reply store-client/pkg/client/mongodb_pipeline_builder.go:214]: we can check healtheventstatus.userpodsevictionstatus directly because node-drainer is updating whole status structure which has both status and message

File: store-client/pkg/client/mongodb_pipeline_builder.go [hunk 3]
```diff
@@ -144,34 +144,102 @@ func (b *MongoDBPipelineBuilder) BuildProcessableNonFatalUnhealthyInsertsPipelin
...
 					),
-					// Watch for unquarantine events (for annotation cleanup)
-					datastore.D(
-						datastore.E("fullDocument.healtheventstatus.nodequarantined", string(model.UnQuarantined)),
-						datastore.E("fullDocument.healtheventstatus.userpodsevictionstatus.status", string(model.StatusSucceeded)),
-					),
-					// Watch for cancelled quarantine events (for annotation cleanup)
+
+					// ============================================================
+					// Case 2: INSERT operations (FAILSAFE PATH)
```
Comment (lalitadithya) [inline store-client/pkg/client/mongodb_pipeline_builder.go:218]: Under what circumstances would this failsafe be needed?

Comment (tanishagoyal2) [reply store-client/pkg/client/mongodb_pipeline_builder.go:218]: Before my change also, pipeline was watching on insertion of event. If i am not wrong @dims wrote that code. @dims Would you please clarify the reason to watch for insertion of event in mongodb and postgres as well.

Comment (dims) [reply store-client/pkg/client/mongodb_pipeline_builder.go:218]: Trying to page this back in.. 

The UPDATE branch is the primary path and is intentionally strict, so it only triggers when the relevant status fields actually changed and avoids duplicate remediation from unrelated updates. The INSERT branch is a safety net for rare pre-populated inserts, such as a migration/backfill job inserting historical events already marked nodequarantined=Quarantined and userpodsevictionstatus.status=Succeeded, a DR replay script re-inserting fully populated documents, a manual DB repair insert of a missing event in final state, or a custom writer/upsert path that creates a new document already in ready/cancelled status (according to codex!). I think i had to do this for some test failures. In normal NVSentinel flow, events are inserted first and statuses are updated later, so this failsafe should fire infrequently but prevents those edge-case events from being missed.

**Review by nitz2407 — APPROVED** [review]

**Review by lalitadithya — APPROVED** [review]
